python/pyspark/ml/functions.py

```diff
@@ -106,6 +117,556 @@ def array_to_vector(col: Column) -> Column:
+            for index, shape in input_tensor_shapes.items():
+                input_shapes[index] = shape
+        else:
+            input_shapes = input_tensor_shapes  # type: ignore
+
+        # iterate over pandas batch, invoking predict_fn with ndarrays
+        for pandas_batch in data:
+            has_tuple = isinstance(pandas_batch, Tuple)  # type: ignore
+            has_tensors = _has_tensor_cols(pandas_batch)
+
+            # require input_tensor_shapes for any tensor columns
+            if has_tensors and not input_shapes:
+                raise ValueError("Tensor columns require input_tensor_shapes")
+
+            for batch in _batched(pandas_batch, batch_size):
+                num_input_rows = len(batch)
+                num_input_cols = len(batch.columns)
+                if num_input_cols == num_expected_cols and num_expected_cols > 1:
+                    # input column per expected input, convert each column into param
+                    multi_inputs = [batch[col].to_numpy() for col in batch.columns]
+                    if input_shapes:
+                        if len(input_shapes) == num_input_cols:
+                            multi_inputs = [
+                                np.vstack(v).reshape([-1] + input_shapes[i])  # type: ignore
+                                if input_shapes[i]
+                                else v
+                                for i, v in enumerate(multi_inputs)
+                            ]
+                            if not all([len(x) == len(batch) for x in multi_inputs]):
+                                raise ValueError("Input data does not match expected shape.")
+                        else:
+                            raise ValueError("input_tensor_shapes must match columns")
+
+                    # run model prediction function on transformed (numpy) inputs
+                    preds = predict_fn(*multi_inputs)
+                elif num_expected_cols == 1:
+                    # multiple input columns for single expected input
+                    if has_tensors:
+                        # tensor columns
+                        if len(batch.columns) == 1:
+                            # one tensor column and one expected input, vstack rows
+                            single_input = np.vstack(batch.iloc[:, 0])
+                        else:
+                            raise ValueError(
+                                "Multiple input columns found, but model expected a single "
+                                "input, use `struct` or `array` to combine columns into tensors."
+                            )
+                    else:
+                        # scalar columns
+                        if len(batch.columns) == 1:
+                            # single scalar column, remove extra dim
+                            single_input = np.squeeze(batch.to_numpy())
+                            if input_shapes and input_shapes[0] not in [None, [], [1]]:
```

Comment:
Issue: 

* If `input_shapes[0] == [1]`, seemingly it will cause issue, it will reshape the values as shape of [-1, 1], and result array becomes 2 dimensions.

Reply:
I think we should restrict the shape to be either None or [] here.

Reply:
Some models require the `[1]` case.  Also note that this is for the input to the model, so any restrictions/limitations from the Spark side should not apply to the model.  (Also refactored the input validation/transform code into separate functions to clean up the core function and make the input vs. output more clear).